File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.java
import java.util.Deque;
import java.util.LinkedList;
import java.util.Scanner;

public class Main {
	static int[] check = new int[26];
	static char[] dna = { 0, 2, 6, 19 };
	static int n, m, ans;
	static String str;
	static Deque<Character> dq = new LinkedList<>();

	static boolean checkDna() {
		for (int i = 0; i < 4; i++)	if (check[dna[i]] > 0)return false;
		return true;
	}
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		n = scanner.nextInt();
		m = scanner.nextInt();
		scanner.nextLine();
		str = scanner.next();
		for (int i = 0; i < 4; i++)	check[dna[i]] = scanner.nextInt();
		for (int i = 0; i < m; i++) {
			check[str.charAt(i) - 'A']--;
			dq.addLast(str.charAt(i));
		}
		if (checkDna())	ans++;
		for (int i = m; i < n; i++) {
            char c = str.charAt(i);
			check[dq.getFirst()-'A']++;
			dq.removeFirst();
			dq.addLast(c);
			check[c - 'A']--;
			if (checkDna())	ans++;
		}
		System.out.println(ans);
	}
}

File: 백준/Silver/12891. DNA 비밀번호/DNA 비밀번호.cc
#include<iostream>
#include<string>
#include<cstring>
#include<deque>
using namespace std;
int check[26];
int dna[4] = { 0,2,6,19 };
int n, m, ans;
string str;
deque<char>sb;
bool checkDna() {
	for (int i = 0; i < 4; i++)	if (check[dna[i]] > 0)	return false;
	return true;
}
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0), cout.tie(0);
	cin >> n >> m;
	cin >> str;
	for (int i = 0; i < 4; i++)cin >> check[dna[i]];
	for (int i = 0; i < m; i++) {
		char c = str[i];
		check[c - 'A']--;
		sb.push_back(str[i]);
	}
	if (checkDna())ans++;
	for (int i = m; i < n; i++) {
		check[sb[0] - 'A']++;
		sb.pop_front();
		sb.push_back(str[i]);
		check[str[i] - 'A']--;
		if (checkDna())ans++;
	}
	cout << ans << "\n";
	return 0;
}